### PR TITLE
Ensure paths stored in manifest are relative

### PR DESCRIPTION
**Pull request by ericmj**

### CODE REVIEW

**Comment by josevalim** [comment]
We should probably do this when we build the records and not as a post
processing step later on.

On Friday, June 3, 2016, Eric Meadows-Jönsson <notifications@github.com>
wrote:

> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/elixir-lang/elixir/pull/4767
> Commit Summary
>
>    - Ensure paths stored in manifest are relative
>
> File Changes
>
>    - *M* lib/mix/lib/mix/compilers/elixir.ex
>    <https://github.com/elixir-lang/elixir/pull/4767/files#diff-0> (12)
>
> Patch Links:
>
>    - https://github.com/elixir-lang/elixir/pull/4767.patch
>    - https://github.com/elixir-lang/elixir/pull/4767.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4767>, or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbsM8_xjVgDsJ7NwYKikCgpV8fH2jks5qH_xUgaJpZM4ItYo6>
> .
>


-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by ericmj** [comment]
@josevalim We already seem to do that (in 1.2.5 also). But that raises the question where the absolute paths are coming from.

**Comment by josevalim** [comment]
@ericmj so we should likely answer that question before merging this. :)

**Comment by ericmj** [comment]
Found the root cause; for dependencies the relative path does not work since `_build/` is relative from the project directory - not from `deps/my_dep`.

The fix is to only store the filename in the manifest and build the absolute path when reading the manifest file.

File: lib/mix/lib/mix/compilers/elixir.ex [hunk 1]
```diff
@@ -91,34 +91,39 @@ defmodule Mix.Compilers.Elixir do
...
   """
-  def read_manifest(manifest) do
+  def read_manifest(manifest, compile_path) do
     # TODO: No longer support file consult once v1.3 is out
     try do
       manifest |> File.read!() |> :erlang.binary_to_term()
     else
-      [@manifest_vsn | t] -> t
-      _ -> []
+      # TODO
```
Comment (josevalim) [inline lib/mix/lib/mix/compilers/elixir.ex]: There is a TODO!

Comment (josevalim) [inline lib/mix/lib/mix/compilers/elixir.ex]: That you didn't do!

File: lib/mix/lib/mix/compilers/elixir.ex [hunk 2]
```diff
@@ -351,37 +351,52 @@ defmodule Mix.Compilers.Elixir do
...
-  defp write_manifest(manifest, modules, sources) do
+  defp write_manifest(manifest, modules, sources, compile_path) do
     File.mkdir_p!(Path.dirname(manifest))
 
     modules =
       for module(beam: beam, binary: binary) = module <- modules do
-        if binary, do: File.write!(beam, binary)
+        if binary do
+          Path.join(compile_path, beam)
+          |> File.write!(binary)
```
Comment (josevalim) [inline lib/mix/lib/mix/compilers/elixir.ex]: You shall never pipe once!

File: lib/mix/lib/mix/tasks/compile.elixir.ex [hunk 3]
```diff
@@ -71,6 +71,8 @@ defmodule Mix.Tasks.Compile.Elixir do
   Cleans up compilation artifacts.
   """
   def clean do
-    Mix.Compilers.Elixir.clean(manifest())
+    project = Mix.Project.config
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.elixir.ex]: No need to call Mix.Project.config if we are using it only once.

File: lib/mix/lib/mix/compilers/elixir.ex (hunk 1 above)
Comment (ericmj) [inline lib/mix/lib/mix/compilers/elixir.ex]: Oops, that's embarrassing.

File: lib/mix/lib/mix/compilers/elixir.ex (hunk 2 above)
Comment (ericmj) [inline lib/mix/lib/mix/compilers/elixir.ex]: Is that the first law of pipe club?

**Comment by ericmj** [comment]
Pushed with fixes based on your comments.

**Comment by josevalim** [comment]
:shipit:

**Pull request merged**